### PR TITLE
[NPM] Always enable connection protocol map cleaner

### PR DESCRIPTION
**Pull request by akarpz**
### What does this PR do?

Always runs the map cleaner for the `connection_protocol` map even when USM is disabled, as it is used by NPM and can be leaky. Also supports USM enabled and NPM disabled.

### Motivation

https://datadoghq.atlassian.net/browse/NPM-3695

### CODE REVIEW

File: pkg/network/tracer/tracer.go [hunk 1]
```diff
@@ -182,6 +185,13 @@ func newTracer(cfg *config.Config, telemetryComponent telemetryComponent.Compone
 	tr.reverseDNS = newReverseDNS(cfg, telemetryComponent)
 	tr.usmMonitor = newUSMMonitor(cfg, tr.ebpfTracer)
 
+	connectionProtocolMap := tr.ebpfTracer.GetMap(probes.ConnectionProtocolMap)
```
Comment (hmahmood) [inline pkg/network/tracer/tracer.go:188]: Should this only be enabled if protocol classification is enabled?

Comment (akarpz) [reply pkg/network/tracer/tracer.go:188]: yes sorry I put this back in draft, it needs a bit more work

File: pkg/network/usm/ebpf_main.go [hunk 2]
```diff
@@ -90,8 +90,8 @@ type ebpfProgram struct {
 	disabledProtocols []*protocols.ProtocolSpec
 
 	// Used for connection_protocol data expiration
-	mapCleaner *ddebpf.MapCleaner[netebpf.ConnTuple, netebpf.ProtocolStackWrapper]
-	buildMode  buildmode.Type
+	connectionProtocolMapCleaner *ddebpf.MapCleaner[netebpf.ConnTuple, netebpf.ProtocolStackWrapper]
```
Comment (guyarb) [inline pkg/network/usm/ebpf_main.go:93]: redundant change

File: pkg/network/tracer/tracer.go [hunk 3]
```diff
@@ -182,6 +185,19 @@ func newTracer(cfg *config.Config, telemetryComponent telemetryComponent.Compone
 	tr.reverseDNS = newReverseDNS(cfg, telemetryComponent)
 	tr.usmMonitor = newUSMMonitor(cfg, tr.ebpfTracer)
 
+	// Setup the connection_protocol map cleaner if protocol classification is enabled and USM is disabled
+	if cfg.ProtocolClassificationEnabled && !usmconfig.IsUSMSupportedAndEnabled(cfg) {
+		connectionProtocolMap := tr.ebpfTracer.GetMap(probes.ConnectionProtocolMap)
+		if connectionProtocolMap == nil {
+			log.Warn("could not get connection_protocol map, will not be able to expire connection_protocol data")
```
Comment (guyarb) [inline pkg/network/tracer/tracer.go:192]: you are logging twice, once in GetMap and once here

Comment (guyarb) [inline pkg/network/tracer/tracer.go:192]:
```suggestion
		connectionProtocolMap := tr.ebpfTracer.GetMap(probes.ConnectionProtocolMap)
		if connectionProtocolMap == nil {
			log.Warnf("could not get %q map, will not be able to expire connection_protocol data", probes.ConnectionProtocolMap)
```

File: pkg/network/usm/ebpf_main.go [hunk 4]
```diff
@@ -488,8 +489,8 @@ func (e *ebpfProgram) init(buf bytecode.AssetReader, options manager.Options) er
 const connProtoTTL = 3 * time.Minute
 const connProtoCleaningInterval = 5 * time.Minute
 
-func (e *ebpfProgram) setupMapCleaner() (*ddebpf.MapCleaner[netebpf.ConnTuple, netebpf.ProtocolStackWrapper], error) {
-	mapCleaner, err := ddebpf.NewMapCleaner[netebpf.ConnTuple, netebpf.ProtocolStackWrapper](e.connectionProtocolMap, 1024)
+func SetupConnectionProtocolMapCleaner(connectionProtocolMap *ebpf.Map) (*ddebpf.MapCleaner[netebpf.ConnTuple, netebpf.ProtocolStackWrapper], error) {
```
Comment (guyarb) [inline pkg/network/usm/ebpf_main.go:492]: exported functions should be documented

File: pkg/network/tracer/tracer.go [hunk 5]
```diff
@@ -182,6 +185,19 @@ func newTracer(cfg *config.Config, telemetryComponent telemetryComponent.Compone
 	tr.reverseDNS = newReverseDNS(cfg, telemetryComponent)
 	tr.usmMonitor = newUSMMonitor(cfg, tr.ebpfTracer)
 
+	// Setup the connection_protocol map cleaner if protocol classification is enabled and USM is disabled
+	if cfg.ProtocolClassificationEnabled && !usmconfig.IsUSMSupportedAndEnabled(cfg) {
```
Comment (hmahmood) [inline pkg/network/tracer/tracer.go:189]: Should this be `cfg.ProtocolClassificationEnabled || usmconfig.IsUSMSupportedAndEnabled(cfg)` and then we don't use the map in the usm monitor/probe?

File: pkg/network/tracer/tracer.go (hunk 3 above)
Comment (akarpz) [reply pkg/network/tracer/tracer.go:192]: I'll remove this log thanks, we want the underlying log from the ebpf tracer

File: pkg/network/tracer/tracer.go (hunk 5 above)
Comment (guyarb) [reply pkg/network/tracer/tracer.go:189]: About initializing it only in npm, it might cause issues in our UTs, as we don't run the tracer, but only USM

Comment (akarpz) [reply pkg/network/tracer/tracer.go]: let me try to fix the UTs

Comment (akarpz) [reply pkg/network/tracer/tracer.go]: update: this didn't actually break any UTs

**Review by hmahmood — APPROVED** [review]

File: pkg/network/tracer/tracer.go (hunk 5 above)
Comment (guyarb) [reply pkg/network/tracer/tracer.go]: it can create flaky tests as the map leaks entries

**Review by guyarb — CHANGES_REQUESTED** [review]
The current change is breaking USM (https://github.com/DataDog/datadog-agent/pull/30194#discussion_r1812241982)

File: pkg/network/tracer/tracer.go (hunk 5 above)
Comment (guyarb) [reply pkg/network/tracer/tracer.go]: if you run full system-probe the code is fine as is today
but, the tests are prune to errors, as we run only usm in the tests (without initializing network_tracer, and hence, without the map cleaner).

**Review by guyarb — APPROVED** [review]

**Comment by akarpz** [comment]
Discussed offline that is should only be an issue in UTs, which shouldn't need the map cleaner anyway

File: pkg/network/tracer/tracer.go (hunk 5 above)
Comment (guyarb) [reply pkg/network/tracer/tracer.go]: Reviewed again - the code itself works as expected.
As for tests, it is unlikely to help (or cause any issue), as the map cleaner runs every 3m, and deletes connections longer than 5m. There is no test today that lasts long enough to be affected by the cleaner.

**Pull request merged**